File: vlc-android/src/org/videolan/vlc/gui/preferences/PreferencesExtensionFragment.java
package org.videolan.vlc.gui.preferences;


import android.content.SharedPreferences;
import android.os.Bundle;
import com.google.android.material.appbar.AppBarLayout;
import androidx.preference.CheckBoxPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceScreen;
import androidx.preference.SwitchPreferenceCompat;
import androidx.preference.TwoStatePreference;

import org.videolan.vlc.R;
import org.videolan.vlc.extensions.ExtensionListing;
import org.videolan.vlc.extensions.ExtensionsManager;
import org.videolan.vlc.util.Settings;

import java.util.ArrayList;
import java.util.List;

public class PreferencesExtensionFragment extends BasePreferenceFragment {

    private ExtensionListing mExtension;
    private int mExtensionId;
    private String mExtensionTitle;
    private String mExtensionKey;
    private String mExtensionPackageName;
    private SharedPreferences mSettings;
    private PreferenceScreen preferenceScreen;
    private boolean androidAutoAvailable = false;
    private List<Preference> preferences = new ArrayList<>();

    @Override
    public void onCreate(Bundle bundle) {
        super.onCreate(bundle);
        mSettings = Settings.INSTANCE.getInstance(requireActivity());
        setHasOptionsMenu(true);
        if (bundle == null)
            bundle = getArguments();

        if (bundle != null) {
            mExtensionKey = bundle.getString("extension_key");
            mExtensionPackageName = mExtensionKey.replace(ExtensionsManager.EXTENSION_PREFIX + "_", "");
            mExtensionId = ExtensionsManager.getInstance().getExtensionId(mExtensionPackageName);
            mExtension = ExtensionsManager.getInstance().getExtensions(getActivity().getApplication(), false).get(mExtensionId);
            mExtensionTitle = mExtension.title();
            setTitle(mExtensionTitle);
            androidAutoAvailable = ExtensionsManager.androidAutoInstalled && mExtension.androidAutoEnabled();
            createCheckboxes();
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putString("extension_key", mExtensionKey);
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onStart() {
        super.onStart();
        ((AppBarLayout) ((PreferencesActivity)getActivity()).findViewById(R.id.appbar)).setExpanded(true, false);
    }

    @Override
    protected int getXml() {
        return R.xml.preferences_extension_page;
    }

    @Override
    protected int getTitleId() {
        return 0;
    }

    private void setTitle(String title){
        final PreferencesActivity activity = (PreferencesActivity)getActivity();
        if (activity != null && activity.getSupportActionBar() != null)
            ((PreferencesActivity)getActivity()).getSupportActionBar().setTitle(title);
    }

    private void createCheckboxes() {
        preferenceScreen = this.getPreferenceScreen();

        //Main switch
        SwitchPreferenceCompat switchPreference = new SwitchPreferenceCompat(preferenceScreen.getContext());
        switchPreference.setTitle(preferenceScreen.getContext().getString(R.string.extension_prefs_activation_title).toUpperCase());
        switchPreference.setKey(mExtensionKey);
        switchPreference.setChecked(mSettings.getBoolean(mExtensionKey, false));
        switchPreference.setOnPreferenceChangeListener(null);
        preferenceScreen.addPreference(switchPreference);

        //Android-auto
        if (androidAutoAvailable) {
            CheckBoxPreference checkbox = new CheckBoxPreference(preferenceScreen.getContext());
            checkbox.setTitle(R.string.android_auto);
            String key = mExtensionKey + "_" + ExtensionsManager.ANDROID_AUTO_SUFFIX;
            checkbox.setKey(key);
            checkbox.setChecked(switchPreference.isChecked() && mSettings.getBoolean(key, false));
            checkbox.setEnabled(switchPreference.isChecked());
            preferences.add(checkbox);
            preferenceScreen.addPreference(checkbox);
        }
    }

    @Override
    public boolean onPreferenceTreeClick(Preference preference) {
        String key = preference.getKey();
        if (key == null || !key.startsWith(mExtensionKey))
            return false;

        if (key.equals(mExtensionKey)) {
            SwitchPreferenceCompat switchPreference = (SwitchPreferenceCompat)preference;
            mSettings.edit().putBoolean(key, switchPreference.isChecked()).apply();
            if (switchPreference.isChecked()) {
                for (Preference checkbox : preferences)
                    checkbox.setEnabled(true);
            } else {
                for (Preference checkbox : preferences) {
                    ((CheckBoxPreference)checkbox).setChecked(false);
                    mSettings.edit().putBoolean(checkbox.getKey(), false).apply();
                    checkbox.setEnabled(false);
                }
            }
        } else if (key.endsWith("_" + ExtensionsManager.ANDROID_AUTO_SUFFIX)) {
            mSettings.edit()
                    .putBoolean(preference.getKey(), ((TwoStatePreference) preference).isChecked())
                    .apply();
        }
        return super.onPreferenceTreeClick(preference);
    }
}

File: application/vlc-android/src/org/videolan/vlc/gui/preferences/PreferencesExtensionFragment.kt
package org.videolan.vlc.gui.preferences


import android.content.SharedPreferences
import android.os.Bundle
import android.view.View
import androidx.preference.CheckBoxPreference
import androidx.preference.Preference
import androidx.preference.SwitchPreferenceCompat
import androidx.preference.TwoStatePreference
import com.google.android.material.appbar.AppBarLayout
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.ObsoleteCoroutinesApi
import org.videolan.vlc.R
import org.videolan.vlc.extensions.ExtensionListing
import org.videolan.vlc.extensions.ExtensionsManager
import org.videolan.tools.Settings
import org.videolan.tools.putSingle
import java.util.*

@ObsoleteCoroutinesApi
@ExperimentalCoroutinesApi
class PreferencesExtensionFragment : BasePreferenceFragment() {

    private var extension: ExtensionListing? = null
    private var extensionId: Int = 0
    private var extensionTitle: String? = null
    private var extensionKey: String? = null
    private var extensionPackageName: String? = null
    private lateinit var settings: SharedPreferences
    // private var preferenceScreen: PreferenceScreen? = null
    private var androidAutoAvailable = false
    private val preferences = ArrayList<Preference>()

    override fun onCreate(bundle: Bundle?) {
        var newBundle = bundle
        super.onCreate(newBundle)
        settings = Settings.getInstance(requireActivity())
        setHasOptionsMenu(true)
        if (newBundle == null)
            newBundle = arguments

        if (newBundle != null) {
            extensionKey = newBundle.getString("extension_key")
            extensionPackageName = extensionKey!!.replace(ExtensionsManager.EXTENSION_PREFIX + "_", "")
            extensionId = ExtensionsManager.getInstance().getExtensionId(extensionPackageName)
            extension = ExtensionsManager.getInstance().getExtensions(activity!!.application, false)[extensionId]
            extensionTitle = extension!!.title()
            setTitle(extensionTitle)
            androidAutoAvailable = ExtensionsManager.androidAutoInstalled && extension!!.androidAutoEnabled()
            createCheckboxes()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putString("extension_key", extensionKey)
        super.onSaveInstanceState(outState)
    }

    override fun onStart() {
        super.onStart()
        ((activity as PreferencesActivity).findViewById<View>(R.id.appbar) as AppBarLayout).setExpanded(true, false)
    }

    override fun getXml(): Int {
        return R.xml.preferences_extension_page
    }

    override fun getTitleId(): Int {
        return 0
    }

    private fun setTitle(title: String?) {
        val activity = activity as PreferencesActivity?
        if (activity != null && activity.supportActionBar != null)
            (getActivity() as PreferencesActivity).supportActionBar!!.title = title
    }

    private fun createCheckboxes() {
        preferenceScreen = this.preferenceScreen

        //Main switch
        val switchPreference = SwitchPreferenceCompat(preferenceScreen!!.context)
        switchPreference.title = preferenceScreen!!.context.getString(R.string.extension_prefs_activation_title).toUpperCase()
        switchPreference.key = extensionKey
        switchPreference.isChecked = settings.getBoolean(extensionKey, false)
        switchPreference.onPreferenceChangeListener = null
        preferenceScreen!!.addPreference(switchPreference)

        //Android-auto
        if (androidAutoAvailable) {
            val checkbox = CheckBoxPreference(preferenceScreen!!.context)
            checkbox.setTitle(R.string.android_auto)
            val key = extensionKey + "_" + ExtensionsManager.ANDROID_AUTO_SUFFIX
            checkbox.key = key
            checkbox.isChecked = switchPreference.isChecked && settings.getBoolean(key, false)
            checkbox.isEnabled = switchPreference.isChecked
            preferences.add(checkbox)
            preferenceScreen!!.addPreference(checkbox)
        }
    }

    override fun onPreferenceTreeClick(preference: Preference): Boolean {
        val key = preference.key
        if (key == null || !key.startsWith(extensionKey!!))
            return false

        if (key == extensionKey) {
            val switchPreference = preference as CheckBoxPreference
            settings.putSingle(key, switchPreference.isChecked)
            if (switchPreference.isChecked) {
                for (checkbox in preferences)
                    checkbox.isEnabled = true
            } else {
                for (checkbox in preferences) {
                    (checkbox as CheckBoxPreference).isChecked = false
                    settings.putSingle(checkbox.getKey(), false)
                    checkbox.setEnabled(false)
                }
            }
        } else if (key.endsWith("_" + ExtensionsManager.ANDROID_AUTO_SUFFIX)) {
            settings.putSingle(preference.key, (preference as TwoStatePreference).isChecked)
        }
        return super.onPreferenceTreeClick(preference)
    }
}